C7Engine/AI/Pathing/BFSLandAlgorithm.cs

```diff
@@ -18,23 +18,31 @@ public class BFSLandAlgorithm : PathingAlgorithm
 		//N.B. This should really be static, but we can't put a static method on interfaces, so it isn't.
 		public TilePath PathFrom(Tile start, Tile destination)
 		{
+			if (start == destination) {
+				return TilePath.EmptyPath(destination);
+			}
+
 			//Distance from start to a given tile
 			Dictionary<Tile, int> distances = new Dictionary<Tile, int>();
 			//Keeps track of which tile (value) preceded a tile (key), so we can reconstruct
 			//the tiles once we know the shortest path.
 			Dictionary<Tile, Tile> predecessors = new Dictionary<Tile, Tile>();
 			Queue<Tile> tilesToVisit = new Queue<Tile>();
 			distances[start] = 0;
-			
+
 			foreach(Tile tile in start.GetLandNeighbors()) {
 				distances[tile] = 1;
 				predecessors[tile] = start;
 				tilesToVisit.Enqueue(tile);
 			}
-			
+
 			//The core BFS algorithm.
 			while (tilesToVisit.Count > 0) {
 				Tile current = tilesToVisit.Dequeue();
+				if (current == destination) {
```

Comment:
@QuintillusCFC tagging you here because we might want to make this change to BFS independent of this PR since it would be nice to get in ASAP for other code using pathing

Reply:
What particular change does line 42 make?  I see how 21-23 covers the case where a player orders a unit to go to the tile it's already on, which wasn't something the AI will do.

It looks like line 42 is intended as a shortcut for the case where the current tile and the destination are the same - but not the start tile and destination?  Thinking through the code, is it correct that this case is going to be triggered only when the destination is the immediate neighbor of the start tile?  Otherwise, it would hit lines 51-53 (which we may be able to remove now?).

I'm too tired to look at the full PR tonight; I might also double-check if I missed something when transcribing this from my Algorithms textbook when less tired (would lines 42-44 make 51-53 no longer required?).

Reply:
Also for context, this is a temporary pathing algorithm chosen due to its simplicity, and the only other place it is used is the settler AI (well, and my just-opened PR will use it for the explorer AI).  So fixing bugs in the short term is good, but we shouldn't aim to highly optimize it.  Hence why I want to make sure I understand the goal.

Reply:
without the change on line 42, pathing fails when the start and destination are adjacent tiles, since all of the adjacent tiles are put into `tilesToVisit`, but we only compare destination to the neighbours of each `tile` in `tileToVisit` when `!distances.Keys.Contains(tile)`, which means that we never actually compare start to neighbouring tiles because they are already in `distances` (I think other tiles are missed too, but I only noticed on neighbouring tiles)